ArduPlane/systemid.cpp

```diff
@@ -0,0 +1,294 @@
+        throttle_offset = 0;
+
+        auto *attitude_control = plane.quadplane.attitude_control;
+        attitude_control->bf_feedforward(restore.att_bf_feedforward);
+        attitude_control->rate_bf_roll_sysid(0);
+        attitude_control->rate_bf_pitch_sysid(0);
+        attitude_control->rate_bf_yaw_sysid(0);
+        plane.quadplane.pos_control->set_xy_control_scale_factor(1);
+
+        // re-initialise the XY controller so we take current position as target
+        plane.quadplane.pos_control->init_xy_controller();
+
+        gcs().send_text(MAV_SEVERITY_INFO, "SystemID stopped");
+    }
+}
+
+/*
+  update systemid - needs to be called at main loop rate
+ */
+void AP_SystemID::update()
+{
+    if (!running) {
+        return;
+    }
+    if (chirp_input.completed()) {
+        stop();
+        return;
+    }
+
+    float const last_loop_time_s = AP::scheduler().get_last_loop_time_s();
+
+    waveform_time += last_loop_time_s;
+    waveform_sample = chirp_input.update(waveform_time, waveform_magnitude);
+    waveform_freq_rads = chirp_input.get_frequency_rads();
+
+    auto *attitude_control = plane.quadplane.attitude_control;
+
+    switch (start_axis) {
+        case AxisType::NONE:
+            // not possible, see start()
+            break;
+        case AxisType::INPUT_ROLL:
+            attitude_offset_deg.x = waveform_sample;
+            break;
+        case AxisType::INPUT_PITCH:
+            attitude_offset_deg.y = waveform_sample;
+            break;
+        case AxisType::INPUT_YAW:
+            attitude_offset_deg.z = waveform_sample;
+            break;
+        case AxisType::RECOVER_ROLL:
+            attitude_offset_deg.x = waveform_sample;
+            attitude_control->bf_feedforward(false);
+            break;
+        case AxisType::RECOVER_PITCH:
+            attitude_offset_deg.y = waveform_sample;
+            attitude_control->bf_feedforward(false);
+            break;
+        case AxisType::RECOVER_YAW:
+            attitude_offset_deg.z = waveform_sample;
+            attitude_control->bf_feedforward(false);
+            break;
+        case AxisType::RATE_ROLL:
+            attitude_control->rate_bf_roll_sysid(radians(waveform_sample));
+            break;
+        case AxisType::RATE_PITCH:
+            attitude_control->rate_bf_pitch_sysid(radians(waveform_sample));
+            break;
+        case AxisType::RATE_YAW:
+            attitude_control->rate_bf_yaw_sysid(radians(waveform_sample));
+            break;
+        case AxisType::MIX_ROLL:
+            attitude_control->actuator_roll_sysid(waveform_sample);
+            break;
+        case AxisType::MIX_PITCH:
+            attitude_control->actuator_pitch_sysid(waveform_sample);
+            break;
+        case AxisType::MIX_YAW:
+            attitude_control->actuator_yaw_sysid(waveform_sample);
+            break;
+        case AxisType::MIX_THROTTLE:
+            throttle_offset += waveform_sample;
+            break;
```

Comment:
This is the same in copter, but `+=` looks wrong to me. surely it should just be `=` @bnsgeyer?

Reply:
@IamPete1 you are correct.  because this is done as an offset which is added to the throttle elsewhere, then it should be =.  In Copter, the throttle is pulled into the mode and then modified and thus why it is a += rather than =.

Reply:
done.